File: java-conversion/src/main/java/com/portfolio/validation/PortfolioValidator.java
package com.portfolio.validation;

import com.portfolio.model.ValidationConstants;
import com.portfolio.model.ValidationRequest;
import com.portfolio.model.ValidationResult;

import java.math.BigDecimal;

public class PortfolioValidator {
    
    public ValidationResult validate(ValidationRequest request) {
        return switch (request.getValidationType()) {
            case ID -> validateId(request.getInputValue());
            case ACCOUNT -> validateAccount(request.getInputValue());
            case TYPE -> validateType(request.getInputValue());
            case AMOUNT -> validateAmount(request.getInputValue());
        };
    }
    
    private ValidationResult validateId(String inputValue) {
        if (inputValue == null || inputValue.length() < 8) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_ID,
                ValidationConstants.ErrorMessages.INVALID_ID
            );
        }
        
        if (!inputValue.substring(0, 4).equals(ValidationConstants.Constants.ID_PREFIX)) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_ID,
                ValidationConstants.ErrorMessages.INVALID_ID
            );
        }
        
        String numericPart = inputValue.substring(4, 8);
        if (!isNumeric(numericPart)) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_ID,
                ValidationConstants.ErrorMessages.INVALID_ID
            );
        }
        
        return ValidationResult.success();
    }
    
    private ValidationResult validateAccount(String inputValue) {
        if (inputValue == null || !isNumeric(inputValue) || inputValue.equals("0000000000")) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_ACCT,
                ValidationConstants.ErrorMessages.INVALID_ACCT
            );
        }
        
        return ValidationResult.success();
    }
    
    private ValidationResult validateType(String inputValue) {
        if (inputValue == null) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_TYPE,
                ValidationConstants.ErrorMessages.INVALID_TYPE
            );
        }
        
        String type = inputValue.trim();
        if (!type.equals("STK") && !type.equals("BND") && 
            !type.equals("MMF") && !type.equals("ETF")) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_TYPE,
                ValidationConstants.ErrorMessages.INVALID_TYPE
            );
        }
        
        return ValidationResult.success();
    }
    
    private ValidationResult validateAmount(String inputValue) {
        if (inputValue == null) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_AMT,
                ValidationConstants.ErrorMessages.INVALID_AMT
            );
        }
        
        try {
            BigDecimal amount = new BigDecimal(inputValue.trim());
            
            if (amount.compareTo(ValidationConstants.Constants.MIN_AMOUNT) < 0 ||
                amount.compareTo(ValidationConstants.Constants.MAX_AMOUNT) > 0) {
                return ValidationResult.error(
                    ValidationConstants.ReturnCodes.INVALID_AMT,
                    ValidationConstants.ErrorMessages.INVALID_AMT
                );
            }
            
            return ValidationResult.success();
        } catch (NumberFormatException e) {
            return ValidationResult.error(
                ValidationConstants.ReturnCodes.INVALID_AMT,
                ValidationConstants.ErrorMessages.INVALID_AMT
            );
        }
    }
    
    private boolean isNumeric(String str) {
        if (str == null || str.isEmpty()) {
            return false;
        }
        
        for (char c : str.toCharArray()) {
            if (!Character.isDigit(c)) {
                return false;
            }
        }
        
        return true;
    }
}

File: src/python/portfolio_management/portfolio/portfolio_validator.py
"""Portfolio Validation Subroutine - migrated from PORTVALD.cbl.

Validates portfolio data elements including ID format, account number,
investment type, and amount ranges.
"""

import logging
from decimal import Decimal, InvalidOperation

from portfolio_management.models.validation import (
    ValidationReturnCode,
    ValidationErrorMessage,
    ValidationConstants,
)

logger = logging.getLogger(__name__)

PROGRAM_ID = "PORTVALD"


class ValidationResult:
    def __init__(self, return_code: int = 0, error_msg: str = ""):
        self.return_code = return_code
        self.error_msg = error_msg

    @property
    def is_valid(self) -> bool:
        return self.return_code == ValidationReturnCode.SUCCESS


class PortfolioValidator:
    def validate(self, validate_type: str, input_value: str) -> ValidationResult:
        if validate_type == "I":
            return self._validate_id(input_value)
        elif validate_type == "A":
            return self._validate_account(input_value)
        elif validate_type == "T":
            return self._validate_type(input_value)
        elif validate_type == "M":
            return self._validate_amount(input_value)
        else:
            return ValidationResult(
                ValidationReturnCode.INVALID_ID,
                "Invalid validation type",
            )

    def _validate_id(self, value: str) -> ValidationResult:
        if len(value) < 8:
            return ValidationResult(
                ValidationReturnCode.INVALID_ID,
                ValidationErrorMessage.INVALID_ID,
            )

        if value[:4] != ValidationConstants.ID_PREFIX:
            return ValidationResult(
                ValidationReturnCode.INVALID_ID,
                ValidationErrorMessage.INVALID_ID,
            )

        if not value[4:8].isdigit():
            return ValidationResult(
                ValidationReturnCode.INVALID_ID,
                ValidationErrorMessage.INVALID_ID,
            )

        return ValidationResult(ValidationReturnCode.SUCCESS)

    def _validate_account(self, value: str) -> ValidationResult:
        if not value.strip().isdigit():
            return ValidationResult(
                ValidationReturnCode.INVALID_ACCT,
                ValidationErrorMessage.INVALID_ACCT,
            )

        if value.strip() == "0" * len(value.strip()):
            return ValidationResult(
                ValidationReturnCode.INVALID_ACCT,
                ValidationErrorMessage.INVALID_ACCT,
            )

        return ValidationResult(ValidationReturnCode.SUCCESS)

    def _validate_type(self, value: str) -> ValidationResult:
        if value.strip() not in ValidationConstants.VALID_INVESTMENT_TYPES:
            return ValidationResult(
                ValidationReturnCode.INVALID_TYPE,
                ValidationErrorMessage.INVALID_TYPE,
            )

        return ValidationResult(ValidationReturnCode.SUCCESS)

    def _validate_amount(self, value: str) -> ValidationResult:
        try:
            amount = Decimal(value.strip())
        except InvalidOperation:
            return ValidationResult(
                ValidationReturnCode.INVALID_AMT,
                ValidationErrorMessage.INVALID_AMT,
            )

        if amount < ValidationConstants.MIN_AMOUNT or amount > ValidationConstants.MAX_AMOUNT:
            return ValidationResult(
                ValidationReturnCode.INVALID_AMT,
                ValidationErrorMessage.INVALID_AMT,
            )

        return ValidationResult(ValidationReturnCode.SUCCESS)
